src/wallet/wallet2.cpp

```diff
@@ -8444,13 +8445,17 @@ void wallet2::get_outs(std::vector<std::vector<tools::wallet2::get_outs_entry>>
           bool own_found = false;
           for (const auto &out: ring)
           {
+            if (seen_outputs.count({amount, out}))
+              seen_indices.emplace(out);
+              continue;
```

Comment:
This if statement seems unnecessary to me.

Assume you have 2 known rings to be used as inputs in a tx, and the first ring includes the real output from the second ring as a decoy. This if statement looks like it would prevent the real output from being added to the second ring, which would then cause the function to throw just outside the for loop because `own_found` won't get set to true. The user's likely recourse in that case is to then remove the ring entirely and have their client construct a new one, which is worse than allowing duplicates.

It seems it's ok to let the duplicates slide here and remove the if statement altogether. The assumption with known rings is that the rings have already been seen by the world, i.e. you're not gaining much by removing the duplicates from rings that have already been seen by the world.

Reply:
Adding to this, it seems the algorithm in general is still able to select duplicates because of this logic

> Assume you have rings to be used as inputs in a tx, and the first ring includes the real output from the second ring as a decoy.

I'm not seeing what prevents this, this could use a deeper look.